unit_tests/src/test_rc_switch_user.cpp

```diff
@@ -0,0 +1,204 @@
+    // NOLINTBEGIN(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+    CHECK(conn_ptr->clientUser.userName == alice.name);
+    CHECK(conn_ptr->clientUser.rodsZone == alice.zone);
+    // NOLINTEND(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+
+    // Show that the test user cannot see the administrator's collection or data object.
+    CHECK_FALSE(fs::client::exists(conn, sandbox));
+    CHECK_FALSE(fs::client::exists(conn, data_object));
+
+    // Become the administrator so that the test can clean up properly.
+    REQUIRE(rc_switch_user(conn_ptr, env.rodsUserName, env.rodsZone) == 0);
+    CHECK(std::strcmp(conn_ptr->clientUser.userName, env.rodsUserName) == 0);
+    CHECK(std::strcmp(conn_ptr->clientUser.rodsZone, env.rodsZone) == 0);
+    // NOLINTEND(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+}
+
+TEST_CASE("rc_switch_user enforces API requirements")
+{
+    //
+    // IMPORTANT: This test requires access to a rodsadmin user!
+    //
+
+    load_client_api_plugins();
+
+    rodsEnv env;
+    _getRodsEnv(env);
+
+    irods::experimental::client_connection conn;
+
+    auto* conn_ptr = static_cast<RcComm*>(conn);
+
+    SECTION("null pointers")
+    {
+        REQUIRE(rc_switch_user(conn_ptr, nullptr, nullptr) == SYS_INVALID_INPUT_PARAM);
+        // Become the administrator so that the test can clean up properly.
+        REQUIRE(rc_switch_user(conn_ptr, env.rodsUserName, env.rodsZone) == 0);
+        CHECK(std::strcmp(conn_ptr->clientUser.userName, env.rodsUserName) == 0);
+        CHECK(std::strcmp(conn_ptr->clientUser.rodsZone, env.rodsZone) == 0);
+        // NOLINTEND(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+
+        REQUIRE(rc_switch_user(conn_ptr, nullptr, "") == SYS_INVALID_INPUT_PARAM);
+        // Become the administrator so that the test can clean up properly.
+        REQUIRE(rc_switch_user(conn_ptr, env.rodsUserName, env.rodsZone) == 0);
+        CHECK(std::strcmp(conn_ptr->clientUser.userName, env.rodsUserName) == 0);
+        CHECK(std::strcmp(conn_ptr->clientUser.rodsZone, env.rodsZone) == 0);
+        // NOLINTEND(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+
+        REQUIRE(rc_switch_user(conn_ptr, "", nullptr) == SYS_INVALID_INPUT_PARAM);
+        // Become the administrator so that the test can clean up properly.
+        REQUIRE(rc_switch_user(conn_ptr, env.rodsUserName, env.rodsZone) == 0);
+        CHECK(std::strcmp(conn_ptr->clientUser.userName, env.rodsUserName) == 0);
+        CHECK(std::strcmp(conn_ptr->clientUser.rodsZone, env.rodsZone) == 0);
+        // NOLINTEND(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+    }
+
+    // clang-format off
+    const std::vector<std::pair<const char*, const char*>> bad_inputs{
+        {"", ""},
+        {"alice", ""},
+        {"", "tempZone"},
+        {"_________________________64_byte_string_________________________", ""},
+        {"", "_________________________64_byte_string_________________________"}
```

Comment:
Should we include tests for non-existent users/zones (but are valid inputs) and usernames which include zone names? Or are these equivalent to passing empty strings?

Reply:
Sure. Will add those cases.

Reply:
done